### PR TITLE
Add React component

### PR DESCRIPTION
**Pull request by trentfridey**
## What

This PR adds the `react-component` package, which exports a `<GenomeSpy />` React component. It also adds documentation on usage and an example in the `embed-examples` package.

## Why

This component should make it is easier to get started with GS in an existing React application.

## How

The component is a thin wrapper around the `embed` function. Its only rendered element is the container, whose `ref` is passed to the `embed` function. It has a `spec` prop for initialization, and a `onEmbed` prop that returns the JS `api` object, so the user has access to `updateNamedData`, etc.

## Notes

As it currently stands, the `spec` is not a reactive prop, i.e. if someone decides to update the object passed to `<GenomeSpy/>`, this won't trigger an update of the visualization. They should use the existing API (as returned in `onEmbed`). Not sure if this would be obvious to a React dev, but I don't have an idea yet as to a better compromise.

Hope this makes sense (not sure if I configured the inter-package dependencies properly). Comments welcome.

### CODE REVIEW

**Comment by tuner** [comment]
Thanks, Trent!

I'll have a look at this today or tomorrow. Meanwhile, could you update `package-lock.json` to get tests running: https://github.com/genome-spy/genome-spy/pull/255/checks

File: packages/react-component/src/main.tsx [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    spec: ViewSpecBase;
+    onEmbed: (api: Promise<EmbedResult>) => void; 
+}
+
+export default function GenomeSpy ({ spec, onEmbed }: IGenomeSpyProps) {
+    const embedRef = useRef<HTMLDivElement>(null)
+    const [error, setError] = useState<string | undefined>()
+
+    useEffect(() => {
+        async function embedToDoc(container: HTMLDivElement | null, config: RootSpec) {
```
Comment (tuner) [inline packages/react-component/src/main.tsx:16]: Not embedding to docs. Function name should be changed.

File: packages/doc-embed/package.json [hunk 2]
```diff
@@ -12,6 +12,7 @@
   },
   "dependencies": {
     "@genome-spy/core": "^0.51.0",
+    "@genome-spy/react-component": "^0.51.0",
```
Comment (tuner) [inline packages/doc-embed/package.json:15]: Not necessary (?)

File: packages/embed-examples/src/reactComponent.html [hunk 3]
```diff
@@ -0,0 +1,18 @@
+<!DOCTYPE html>
+<html lang="en">
+
+<head>
+    <meta charset="UTF-8" />
+    <link rel="stylesheet" href="style.scss" />
+    <title>React Component example</title>
+    <script type="module" src="/reactComponent.jsx"></script>
```
Comment (tuner) [inline packages/embed-examples/src/reactComponent.html:8]: Should be at the end of body. To make it consistent with other examples.

File: packages/embed-examples/tsconfig.json [hunk 4]
```diff
@@ -1,3 +1,6 @@
 {
+  "compilerOptions": {
+    "jsx": "react"
```
Comment (tuner) [inline packages/embed-examples/tsconfig.json:3]: A path entry to `packages/react-component/src` could be added here to fix typings. See the `/tsconfig.json` at the repo root.

File: packages/react-component/src/main.tsx [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import type { ViewSpecBase, RootSpec } from "@genome-spy/core"
+import { EmbedResult } from '@genome-spy/core/types/embedApi.js';
+
+interface IGenomeSpyProps {
+    spec: ViewSpecBase;
+    onEmbed: (api: Promise<EmbedResult>) => void; 
+}
+
+export default function GenomeSpy ({ spec, onEmbed }: IGenomeSpyProps) {
+    const embedRef = useRef<HTMLDivElement>(null)
```
Comment (tuner) [inline packages/react-component/src/main.tsx:12]: Could `containerRef` be more specific?

File: packages/react-component/src/main.tsx [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    const [error, setError] = useState<string | undefined>()
+
+    useEffect(() => {
+        async function embedToDoc(container: HTMLDivElement | null, config: RootSpec) {
+            try {
+               const api = await embed(container, config, { bare: true })
```
Comment (tuner) [inline packages/react-component/src/main.tsx:18]: If the component is removed from the dom, embed api's `finalize` method should be called to ensure that all resources are released, possible listeners in the document root are removed, etc.

File: packages/embed-examples/src/reactComponent.jsx [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+                        field: "y",
+                        type: "quantitative",
+                        scale: { domain: [-1, 1] },
+                    },
+                },
+            },
+        ],
+    }
+
+    const generateData = () => {
```
Comment (tuner) [inline packages/embed-examples/src/reactComponent.jsx:33]: This is overly complex for a simple example with static data. I would provide a flat, static array of objects.

**Review by tuner — COMMENTED** [review]
Seems to work! I added some comments.

When it comes to changing the spec, etc., could we implement it in pretty much the same way as `react-vega` does? https://vega.github.io/react-vega/?path=/story/react-vega--vegalite

Thus, when the identity of the provided spec object changes, i.e., the user provides a new spec, the component destroys the existing GenomeSpy instance and calls `embed` function again with the new spec?

It would be convenient to allow dynamic data to be updated in a similar way. When the user provides a new data object (which maps dataset names to data arrays) through the `data` attribute, those dataset arrays that have a new identity would be updated using `updateNamedData`.

File: packages/embed-examples/src/reactComponent.jsx (hunk 7 above)
Comment (trentfridey) [reply packages/embed-examples/src/reactComponent.jsx]: Should be fixed in 0ba4f7a

File: packages/react-component/src/main.tsx (hunk 6 above)
Comment (trentfridey) [reply packages/react-component/src/main.tsx]: Thanks, should be in 093b914

File: packages/react-component/src/main.tsx (hunk 1 above)
Comment (trentfridey) [reply packages/react-component/src/main.tsx]: Thanks, should be in 093b914

File: packages/react-component/src/main.tsx (hunk 5 above)
Comment (trentfridey) [reply packages/react-component/src/main.tsx]: Thanks, should be in 093b914

File: packages/embed-examples/tsconfig.json (hunk 4 above)
Comment (trentfridey) [reply packages/embed-examples/tsconfig.json]: Added in https://github.com/genome-spy/genome-spy/pull/255/commits/1abffe2dba59b1c455dbe97aab1276efc23b57f6

File: packages/embed-examples/src/reactComponent.html (hunk 3 above)
Comment (trentfridey) [reply packages/embed-examples/src/reactComponent.html]: Fixed in https://github.com/genome-spy/genome-spy/pull/255/commits/a5144241529694463dd8ae9b0405b551b2e38228

**Comment by tuner** [comment]
Thanks, Trent. I'll have a look at this later this week.

**Pull request merged**